File: algo/src/test/java/com/hartwig/actin/algo/evaluation/toxicity/HasIntoleranceRelatedToStudyMedicationTest.java
package com.hartwig.actin.algo.evaluation.toxicity;

import static com.hartwig.actin.algo.evaluation.EvaluationAssert.assertEvaluation;

import com.google.common.collect.Lists;
import com.hartwig.actin.algo.datamodel.EvaluationResult;
import com.hartwig.actin.clinical.datamodel.Intolerance;

import org.junit.Test;

public class HasIntoleranceRelatedToStudyMedicationTest {

    @Test
    public void canEvaluate() {
        HasIntoleranceRelatedToStudyMedication function = new HasIntoleranceRelatedToStudyMedication();

        // No allergies
        assertEvaluation(EvaluationResult.FAIL, function.evaluate(ToxicityTestFactory.withIntolerances(Lists.newArrayList())));

        // Random allergy
        Intolerance wrongCategory = ToxicityTestFactory.intolerance().category("some category").build();
        assertEvaluation(EvaluationResult.FAIL, function.evaluate(ToxicityTestFactory.withIntolerance(wrongCategory)));

        // Inactive allergy for medication
        Intolerance inactive = ToxicityTestFactory.intolerance().category(HasIntoleranceRelatedToStudyMedication.MEDICATION_CATEGORY).build();
        assertEvaluation(EvaluationResult.FAIL, function.evaluate(ToxicityTestFactory.withIntolerance(inactive)));

        // Actual relevant allergy
        Intolerance relevantCategory = ToxicityTestFactory.intolerance()
                .category(HasIntoleranceRelatedToStudyMedication.MEDICATION_CATEGORY.toUpperCase())
                .clinicalStatus(HasIntoleranceRelatedToStudyMedication.CLINICAL_STATUS_ACTIVE.toUpperCase())
                .build();
        assertEvaluation(EvaluationResult.UNDETERMINED, function.evaluate(ToxicityTestFactory.withIntolerance(relevantCategory)));

        // Actual relevant allergy due to DOID
        Intolerance relevantDoid = ToxicityTestFactory.intolerance()
                .addDoids(HasIntoleranceRelatedToStudyMedication.DRUG_ALLERGY_DOID)
                .clinicalStatus(HasIntoleranceRelatedToStudyMedication.CLINICAL_STATUS_ACTIVE)
                .build();
        assertEvaluation(EvaluationResult.UNDETERMINED, function.evaluate(ToxicityTestFactory.withIntolerance(relevantDoid)));
    }
}

File: algo/src/test/kotlin/com/hartwig/actin/algo/evaluation/toxicity/HasIntoleranceRelatedToStudyMedicationTest.kt
package com.hartwig.actin.algo.evaluation.toxicity

import com.hartwig.actin.algo.evaluation.EvaluationAssert.assertEvaluation
import com.hartwig.actin.algo.evaluation.othercondition.ComorbidityTestFactory
import com.hartwig.actin.algo.icd.IcdConstants
import com.hartwig.actin.datamodel.algo.EvaluationResult
import com.hartwig.actin.datamodel.clinical.ToxicitySource
import com.hartwig.actin.icd.IcdModel
import com.hartwig.actin.icd.datamodel.IcdNode
import org.junit.Test

class HasIntoleranceRelatedToStudyMedicationTest {

    private val icdModel = IcdModel.create(
        listOf(
            IcdNode(
                IcdConstants.NIVOLUMAB_CODE,
                listOf(IcdConstants.ANTI_NEOPLASTIC_AGENTS, IcdConstants.DRUG_ALLERGY_CODE),
                "Nivolumab"
            )
        )
    )
    private val function = HasIntoleranceRelatedToStudyMedication(icdModel)
    private val matchingIcdCodes = IcdConstants.DRUG_ALLERGY_SET

    @Test
    fun `Should fail when no comorbidities in history`() {
        assertEvaluation(EvaluationResult.FAIL, function.evaluate(ComorbidityTestFactory.withIntolerances(emptyList())))
    }

    @Test
    fun `Should fail when intolerance has wrong ICD code`() {
        val intolerance = ComorbidityTestFactory.intolerance(icdMainCode = "wrong")
        assertEvaluation(EvaluationResult.FAIL, function.evaluate(ComorbidityTestFactory.withComorbidity(intolerance)))
    }

    @Test
    fun `Should fail when intolerance with matching ICD code is not active`() {
        val intolerance = ComorbidityTestFactory.intolerance(icdMainCode = matchingIcdCodes.first(), clinicalStatus = "other")
        assertEvaluation(EvaluationResult.FAIL, function.evaluate(ComorbidityTestFactory.withComorbidity(intolerance)))
    }

    @Test
    fun `Should fail when toxicity with matching ICD code has grade below 2 or null`() {
        val icdMainCode = matchingIcdCodes.first()
        listOf(
            ComorbidityTestFactory.toxicity("tox", ToxicitySource.EHR, 1, icdMainCode),
            ComorbidityTestFactory.toxicity("tox", ToxicitySource.EHR, null, icdMainCode)
        ).forEach { match ->
            assertEvaluation(EvaluationResult.FAIL, function.evaluate(ComorbidityTestFactory.withComorbidity(match)))
        }
    }

    @Test
    fun `Should fail when allergy is not to anti-cancer agent`() {
        val intolerance = ComorbidityTestFactory.intolerance(
            icdMainCode = matchingIcdCodes.first(),
            icdExtensionCode = IcdConstants.CEFAMYCIN_ANTIBIOTIC,
            clinicalStatus = HasIntoleranceRelatedToStudyMedication.CLINICAL_STATUS_ACTIVE
        )
        assertEvaluation(EvaluationResult.FAIL, function.evaluate(ComorbidityTestFactory.withComorbidity(intolerance)))
    }

    @Test
    fun `Should evaluate to undetermined when active intolerance has matching ICD code`() {
        val intolerance = ComorbidityTestFactory.intolerance(
            icdMainCode = matchingIcdCodes.first(),
            icdExtensionCode = IcdConstants.NIVOLUMAB_CODE,
            clinicalStatus = HasIntoleranceRelatedToStudyMedication.CLINICAL_STATUS_ACTIVE
        )
        assertEvaluation(EvaluationResult.UNDETERMINED, function.evaluate(ComorbidityTestFactory.withComorbidity(intolerance)))
    }

    @Test
    fun `Should evaluate to undetermined when any other comorbidity has matching ICD code`() {
        val icdMainCode = matchingIcdCodes.first()
        listOf(
            ComorbidityTestFactory.toxicity("tox", ToxicitySource.EHR, 2, icdMainCode, IcdConstants.NIVOLUMAB_CODE),
            ComorbidityTestFactory.otherCondition("condition", icdMainCode = icdMainCode, icdExtensionCode = IcdConstants.NIVOLUMAB_CODE),
            ComorbidityTestFactory.complication("complication", icdMainCode = icdMainCode, icdExtensionCode = IcdConstants.NIVOLUMAB_CODE)
        ).forEach { match ->
            assertEvaluation(EvaluationResult.UNDETERMINED, function.evaluate(ComorbidityTestFactory.withComorbidity(match)))
        }
    }
}